File: 005_longest_palindromic_substring/Solution.py
class Solution:
    def preprocess(self, s: str) -> str:
        result = '^'
        for ele in s:
            result += '#' + ele
        return result + '#$'

    def longestPalindrome(self, s: str) -> str:
        tmp_str = self.preprocess(s)
        str_len = len(tmp_str)
        p = [0] * str_len
        c, r = 0, 0
        index = 0
        max_len = 0

        for i in range(1, str_len - 1):
            p[i] = min(p[2 * c - i], r - i) if r > i else 1

            while tmp_str[i + p[i]] == tmp_str[i - p[i]]:
                p[i] += 1

            if r < i + p[i]:
                r = i + p[i]
                c = i

            if max_len < p[i] - 1:
                max_len = p[i] - 1
                index = i

        start = (index - max_len) // 2
        return s[start: start + max_len]


if __name__ == '__main__':
    solution = Solution()
    print(solution.longestPalindrome('babad'))

File: 005_longest_palindromic_substring/Solution.go
package main

import (
    "fmt"
    "math"
)

func preprocess(s string) string {
    result := "^"
    for _, ele := range s {
        result += "#" + string(ele)
    }

    return result + "#$"
}

func longestPalindrome(s string) string {
    if len(s) == 0 {
        return ""
    }

    tmpStr := preprocess(s)
    strLen := len(tmpStr)
    p := make([]int, strLen)
    C, R := 0, 0
    index, maxLen := 0, 0

    for i := 1; i < strLen-1; i++ {
        if R > i {
            p[i] = int(math.Min(float64(p[2*C-i]), float64(R-i)))
        } else {
            p[i] = 1
        }

        for tmpStr[i+p[i]] == tmpStr[i-p[i]] {
            p[i]++
        }

        if R < i+p[i] {
            R = i + p[i]
            C = i
        }

        if maxLen < p[i]-1 {
            maxLen = p[i] - 1
            index = i
        }
    }

    start := (index - maxLen) / 2
    return s[start: start+maxLen]
}

func main() {
    fmt.Println(longestPalindrome("babad"))
}
